include/oneapi/tbb/task_group.h

```diff
@@ -431,24 +435,122 @@ class structured_task_group;
+        ++m_ref_counter;
+    }
+
+    void remove_ref() {
+        if (--m_ref_counter == 0) {
+            finalize();
+        }
+    }
+
+private:
+    void finalize() {
         // Make a local reference not to access this after destruction.
-        wait_context& wo = m_wait_ctx;
-        // Copy allocator to the stack
+        auto& wo = m_wait_ctx;
+        task_group_continuation* parent = m_parent;
+
         auto allocator = m_allocator;
-        // Destroy user functor before release wait.
-        this->~function_task();
-        wo.release();
 
-        allocator.deallocate(this, ed);
+        this->~task_group_continuation();
+
+        if (parent) {
+            parent->remove_ref();
+        } else {
+            wo.release();
+        }
+
+        allocator.deallocate(this);
+    }
+
+    std::atomic<std::uint64_t> m_ref_counter{0};
+    task_group_continuation* m_parent{nullptr};
+    wait_context& m_wait_ctx;
+    small_object_allocator m_allocator;
+};
+
+class base_task_group_task : public task {
+public:
+    base_task_group_task(wait_context& wo, small_object_allocator& alloc, task_group_continuation* p = nullptr)
+        : m_wait_ctx(wo), m_allocator(alloc), m_parent(p)
+    {}
+
+    task_group_continuation* get_continuation() {
+        if (m_continuation == nullptr) {
+            small_object_allocator alloc{};
+            m_continuation = alloc.new_object<task_group_continuation>(m_parent, m_wait_ctx, alloc);
+            // Original task holds implicit reference
+            m_continuation->add_ref();
+        }
+        return m_continuation;
+    }
+
+    bool is_same_task_group(wait_context* wo) {
+        return &m_wait_ctx == wo;
+    }
+protected:
+    wait_context& m_wait_ctx;
+    small_object_allocator m_allocator;
+    task_group_continuation* m_parent{nullptr};
+    task_group_continuation* m_continuation{nullptr};
+};
+
+template<typename F>
+class function_task : public base_task_group_task {
+    const F m_func;
+
+    void finalize(const execution_data& ed) {
+        if (m_continuation == nullptr) {
+            // Make a local reference not to access this after destruction.
+            wait_context& wo = m_wait_ctx;
+            auto allocator = m_allocator;
+            auto parent = m_parent;
+
+            // Destroy user functor before release wait.
+            this->~function_task();
+
+            if (parent) {
+                parent->remove_ref();
+            } else {
+                wo.release();
+            }
+
+            allocator.deallocate(this, ed);
+        } else {
+            auto continuation = m_continuation;
+            auto allocator = m_allocator;
+
+            // Destroy user functor before release wait.
+            this->~function_task();
+
+            if (continuation) {
+                // Original task holds implicit reference on continuation
+                continuation->remove_ref();
+            }
+
+            allocator.deallocate(this, ed);
```

Comment:
Can we move it 2 lines down not to duplicate same actions for these two branches.